Kerberos authentication via webui and some minor improvements

bgeuken opened the pull request:
This PR implements authentication via kerberos in the webui.

With kerberos authentication enabled, users authenticate via a kerberos ticket to the OBS. In case of authentication via browser, users have to configure their browser to use kerberos auth for the domain of the OBS server.

Users that have not done that or do not have a kerberos ticket will be redirected to a login page that informs them about enabled kerberos mode.

OBS creates a new account, when a users logs in the first time.

In addition it:
* Introduces a kerberos_mode flag
* Cleans up setting of last_logged_in_at
* Removes validation of email presence. Kerberos doesn't provide an email, which would mean that newly created users are invalid.

Pair-programmed with @eduardoj

evanrolfe commented on src/api/spec/models/authenticator_spec.rb:
```diff
@@ -0,0 +1,74 @@
+
+        context 'and confirmed user' do
+          it 'authenticates the user' do
+            authenticator = Authenticator.new(request_mock, session_mock, response_mock)
```
In some parts of this spec the `Authenticator` is instantiated in a let block but in other its instantiated inside an it block, so it would be nice to keep it consistent.

evanrolfe commented on src/api/spec/features/webui/login_spec.rb:
```diff
@@ -91,4 +92,84 @@
+      }))
+    end
+
+    context 'for a normal request that requires login' do
```
"normal request" here is pretty vague, maybe change this line to something like this?
```ruby
context 'a request without the authorization header' do
```

evanrolfe commented on src/api/spec/features/webui/login_spec.rb:
```diff
@@ -91,4 +92,84 @@
+        click_link("Log In")
+      end
+
+      context "and 'Negotiate' header is not set" do
```
May as well just get rid of this line and move the context description to the parent context on line 105

bgeuken replied:
I like it this way

evanrolfe commented on src/api/spec/models/authenticator_spec.rb:
```diff
@@ -0,0 +1,74 @@
+    }))
+  end
+
+  describe 'extract_user' do
```
Should be `describe '#extract_user'` since its an instance method

evanrolfe commented on src/api/spec/requests/kerberos_login_spec.rb:
```diff
@@ -1,82 +1,64 @@
+        before do
+          allow_any_instance_of(Authenticator).to receive(:extract_krb_user).
+            with(auth_header.to_s.split).
+            and_raise(Authenticator::AuthenticationRequiredError, 'something happened')
```
It would be preferable to only mock the external API here rather than another internal class (the Authenticator class). So you could mock GSSApi here like you did earlier in this spec.

bgeuken replied:
Well, the authenticator class get's tested intensively (at least the kerberos related parts...). So this should be fine.

evanrolfe replied:
> the authenticator class get's tested intensively

Thats not the point. Integration tests are there to test the application as a whole without any knowledge of whats going on inside. This means that they give us confidence to refactor code if we want to and also make changes to other parts of the code while knowing that we're not breaking the functionality here. So we shouldn't mock our internal classes here because what if we want to change them without changing the behaviour? 

For example when I moved the authentication code into the `Authenticator` class I relied on the functional tests more than I did on the rspec tests because the functional ones did a better job of testing the application like a black box (only testing the output was correct given a certain input). If it wasn't for those functional tests then I probably wouldn't have had the confidence to refactor the code because it is such an important part of the application.

bgeuken replied:
We do have knowledge about parts of the implementation anyway. We all agreed not to run a kerberos server in the test suite;-) So we have knowledge about the part that we mock. The only thing we can decide on is, if

* we want test each and every possible GSSAPI error in an integration test (and mock each of them)
* we only test one of the GSSAPI cases, mock that one and ignore the others
* we mock the GSSAPI handling of the Authenticator class and fake a failure there

We have unit tests that ensure the different GSSAPI failure cases are correctly handled by the authenticator. So going with the third option allows us to verify if failures are propagated correctly without worrying to much about specific failure types / scenarios. Which IMO isn't the worst for an integration test.

evanrolfe commented on src/api/spec/requests/kerberos_login_spec.rb:
```diff
@@ -1,82 +1,64 @@
 require 'rails_helper'
-
+require 'authenticator'
```
It would be ideal if we could get rid of this require since request specs are supposed to be integration tests so the only things that should be mocked are external services like GSSApi.

bgeuken commented on src/api/lib/authenticator.rb:
```diff
@@ -87,11 +87,27 @@ def require_admin
+        return request.env[header].to_s.split
+      end
+    end
+    return
```
There is an each iterator, so we need the return.

Probably would be nicer to return an array here. But I was just moving the code out of the private section. So I would prefer not to do that here.

bgeuken commented:
@evanrolfe I've updated the code according to your review. In addition I also added a shared context that set's up the gssapi mock (for valid tickets). I squashed the change into b36cdb1c8ba6a38c3ca18a15b139dd8c8f23d0e9

evanrolfe commented on src/api/spec/models/authenticator_spec.rb:
```diff
@@ -0,0 +1,87 @@
+require 'authenticator'
+require 'gssapi'
+
+RSpec.describe Authenticator do
```
Just noticed this but you've created the spec here: `spec/models/authenticator_spec.rb` but `Authenticator` is not a model its in the lib directory so shouldn't the spec go in `spec/lib/authenticator_spec.rb`?

bgeuken commented on ReleaseNotes-2.9:
```diff
@@ -28,9 +28,10 @@ Features
 
 UI:
- *
+ * Kerberos authentication mode
```
@hennevogel Added a short sentence and a pointer to our admin guide.
